src/FunderMaps.Core.Authentication/SignInService.cs

```diff
@@ -0,0 +1,226 @@
+
+        /// <summary>
+        ///     The <see cref="IPasswordHasher"/> used.
+        /// </summary>
+        public IPasswordHasher PasswordHasher { get; }
+
+        /// <summary>
+        ///     Gets the <see cref="ILogger"/> used to log messages.
+        /// </summary>
+        protected ILogger Logger { get; }
+
+        /// <summary>
+        ///     Creates a new instance.
+        /// </summary>
+        public SignInService(
+            IUserRepository userRepository,
+            IOrganizationUserRepository organizationUserRepository,
+            IOrganizationRepository organizationRepository,
+            IPasswordHasher passwordHasher,
+            ILogger<SignInService> logger)
+        {
+            UserRepository = userRepository ?? throw new ArgumentNullException(nameof(userRepository));
+            OrganizationUserRepository = organizationUserRepository ?? throw new ArgumentNullException(nameof(organizationUserRepository));
+            OrganizationRepository = organizationRepository ?? throw new ArgumentNullException(nameof(organizationRepository));
+            PasswordHasher = passwordHasher ?? throw new ArgumentNullException(nameof(passwordHasher));
+            Logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        /// <summary>
+        ///     Returns a flag indicating whether the specified user can sign in.
+        /// </summary>
+        /// <param name="id">The user id whose sign-in status should be returned.</param>
+        /// <returns><c>True</c> if the specified user can sign-in, otherwise false.</returns>
+        public virtual async Task<SignInContext> CanSignInAsync(Guid id)
+        {
+            if (await UserRepository.IsLockedOutAsync(id))
+            {
+                Logger.LogWarning(3, $"User {id} is currently locked out.");
+
+                return SignInContext.LockedOut;
+            }
+
+            return SignInContext.Success;
+        }
+
+        /// <summary>
+        ///     Test if the provided password is valid for the user.
+        /// </summary>
+        /// <param name="id">The user id whose password should be checked.</param>
+        /// <param name="password">The password to be checked against the user.</param>
+        /// <returns><c>True</c> if the specified password is valid for the user, otherwise false.</returns>
+        public virtual async Task<bool> CheckPasswordAsync(Guid id, string password)
+        {
+            var passwordHash = await UserRepository.GetPasswordHashAsync(id);
+            return PasswordHasher.IsPasswordValid(passwordHash, password);
+        }
+
+        /// <summary>
+        ///     Set the password for the user.
+        /// </summary>
+        /// <param name="id">The user id whose password should be set.</param>
+        /// <param name="password">The password to be set on the user.</param>
+        public virtual async Task SetPasswordAsync(Guid id, string password)
+        {
```

Comment:
We should be careful with exposing this method, as there might be situations where we explicitly want to check if whoever calls this method also has the correct **current** user password. For admin executors this should be allowed, but for non-admin executors a current password might be required.

This is worth double checking.

Reply:
We'll leave it be. If we find ambiguity in future development then there is the possibility to shove this into a separate password service